File: ruby/SpaceCity.rb
#encoding-utf:8

require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'

module Deepspace
    class SpaceCity < SpaceStation
        def initialize(bas, rest) #base sera una spacestation y rest un array de spacestation
            copy(bas)
            @base = bas
            @collaborators = rest
        end

        attr_reader :collaborators

        def fire
            dam =  super
            @collaborators.each do |stations|
              dam += stations.fire
            end
            return dam
        end

        def protection
            sh =  super
            @collaborators.each do |stations|
              sh += stations.protection
            end
            return sh
        end

        def setLoot(loot)
            super
            return Transformation::NOTRANSFORM
        end

        def getUIversion
            return SpaceCityToUI.new(self)
        end

        def to_s
            b = "Base: " + @base.to_s
            c = "Colaboradores: "
      
            @collaborators.each { |colab|
              c += colab.to_s + "\n"
            }
            
            return b + c
        end
    end
end

File: java/Deepspace/src/deepspace/SpaceCity.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;
import java.util.ArrayList;

/**
 *
 * @author alvaro
 */
public class SpaceCity extends SpaceStation{
    private SpaceStation base;
    private ArrayList<SpaceStation> collaborators;
    
    SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        
        this.base = base;
        collaborators = new ArrayList<SpaceStation>(rest);        
    }
    
    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public float fire(){
        float result = super.fire();
        
        for(int i=0; i<collaborators.size(); i++){
            result += collaborators.get(i).fire();
        }
        
        return result;
    }
    
    @Override
    public float protection(){
        float result = super.protection();
        
        for(int i=0; i<collaborators.size(); i++){
            result += collaborators.get(i).protection();
        }
        
        return result;
    }
    
    @Override
    public SpaceCityToUI getUIversion(){
        return new SpaceCityToUI(this);
    }
    
    @Override
    public String toString(){
        String bas = "SPACE CITY: \n" + "Base: " + base.toString() +"\n";
        String col = "Collaborators: \n";
        
        for (int i=0; i<collaborators.size();i++){
            col += collaborators.get(i).toString() + "\n";
        }
        
        String result = bas + col;
        
        return result;
    }
    
}
